src/OpenDiffix.Core/Anonymizer.fs

```diff
@@ -47,25 +47,39 @@ let private generateNoise salt stepName stdDev noiseLayers =
 
 // Compacts flattening intervals to fit into the total count of contributors.
 // Both intervals are reduced proportionally, with `topCount` taking priority.
+// NOTE `totalCount >= outlierCount.Lower + topCount.Lower` is assumed
 let private compactFlatteningIntervals outlierCount topCount totalCount =
   let totalAdjustment = outlierCount.Upper + topCount.Upper - totalCount
 
   if totalAdjustment <= 0 then
     outlierCount, topCount // no adjustment needed
   else
+    // NOTE: at this point we know `0 <= totalAdjustment <= outlierRange + topRange` (*)
+    //       because `totalAdjustment = outlierCount.Upper + topCount.Upper - totalCount
+    //                               <= outlierCount.Upper + topCount.Upper - outlierCount.Lower - topCount.Lower`
+    //       by the condition checked in `aidFlattening`
     let outlierRange = outlierCount.Upper - outlierCount.Lower
-
-    if outlierRange > topCount.Upper - topCount.Lower then
-      failwith "Invalid config: OutlierCount interval is larger than TopCount interval."
-
-    let outlierAdjustment, topAdjustment =
-      if outlierRange < totalAdjustment / 2 then
-        outlierRange, totalAdjustment - outlierRange
-      else
-        totalAdjustment / 2, totalAdjustment - totalAdjustment / 2
-
-    { outlierCount with Upper = outlierCount.Upper - outlierAdjustment },
-    { topCount with Upper = topCount.Upper - topAdjustment }
+    let topRange = topCount.Upper - topCount.Lower
+    // `topAdjustment` will be half of `totalAdjustment` rounded up, so it takes priority as it should
+    let outlierAdjustment = totalAdjustment / 2
+    let topAdjustment = totalAdjustment - outlierAdjustment
+
+    if outlierRange >= outlierAdjustment && topRange >= topAdjustment then
+      // both ranges are compacted at same rate
+      { outlierCount with Upper = outlierCount.Upper - outlierAdjustment },
+      { topCount with Upper = topCount.Upper - topAdjustment }
+    elif outlierRange < outlierAdjustment then
+      // `outlierCount` is compacted as much as possible by `outlierRange`, `topCount` takes the surplus adjustment
+      { outlierCount with Upper = outlierCount.Lower },
+      { topCount with Upper = topCount.Upper - totalAdjustment + outlierRange }
+    elif topRange < topAdjustment then
+      // vice versa
+      { outlierCount with Upper = outlierCount.Upper - totalAdjustment + topRange },
+      { topCount with Upper = topCount.Lower }
```

Comment:
I'm finding it difficult to figure out what's supposed to happen. Maybe Cristian could take a look as well.

Reply:
I think some unit tests would be useful to ensure correctness here.

Reply:
Otherwise, looks correct to me.

Reply:
+1 unit tests, thanks for the nudge. I was considering splitting the compaction to a tiny dedicated module, and making a low-level test for that, but not sure if not overkill.

Otherwise the unit test would be very tailor-made to enact the different paths. WDYT?

Reply:
I wouldn't make a separate module just for this. It leads to overly-fragmented code.
Marking the function `public` so it can be tested would be best in my opinion.